File: dotnet/src/MissionControl.Client/AutoCancelOrderHandler.cs
using System.Text;
using System.Text.Json;

namespace MissionControl.Client;

/// <summary>
/// A <see cref="DelegatingHandler"/> that automatically attempts to cancel an order when a
/// request to one of the order-mutation endpoints fails with a server error (5xx) or a
/// connection/timeout error.
/// <para>
/// Affected endpoints: <c>createOrder</c>, <c>claimReservation</c>, <c>reportOrder</c>.
/// </para>
/// <para>
/// The cancellation attempt is best-effort. If it fails the original error is still propagated
/// to the caller unchanged.
/// </para>
/// </summary>
internal sealed class AutoCancelOrderHandler : DelegatingHandler
{
    /// <summary>
    /// Maximum time to wait for the compensating cancel request to complete.
    /// </summary>
    private static readonly TimeSpan CancelTimeout = TimeSpan.FromSeconds(30);


    /// <inheritdoc/>
    protected override async Task<HttpResponseMessage> SendAsync(
        HttpRequestMessage request,
        CancellationToken cancellationToken)
    {
        if (!IsAutoCancelEndpoint(request))
            return await base.SendAsync(request, cancellationToken).ConfigureAwait(false);

        // Buffer the request body before SendAsync consumes the stream.
        // ReadAsStringAsync calls LoadIntoBufferAsync internally, so the content remains
        // sendable by the inner handlers.
        string? requestBody = null;
        if (request.Content is not null)
            requestBody = await request.Content.ReadAsStringAsync(cancellationToken).ConfigureAwait(false);

        HttpResponseMessage response;
        try
        {
            response = await base.SendAsync(request, cancellationToken).ConfigureAwait(false);
        }
        catch (Exception ex) when (ex is HttpRequestException or TaskCanceledException)
        {
            // Connection error or timeout — attempt to cancel with a fresh token so that
            // the cancel request is not affected by the original token's cancellation.
            await TryCancelOrderAsync(request, requestBody).ConfigureAwait(false);
            throw;
        }

        if (IsServerError(response))
            await TryCancelOrderAsync(request, requestBody).ConfigureAwait(false);

        return response;
    }

    private static bool IsAutoCancelEndpoint(HttpRequestMessage request)
    {
        var path = request.RequestUri?.AbsolutePath;
        if (path is null)
            return false;

        // Use Contains so this works regardless of any base-path prefix
        // (e.g. /partnerApi/v2/Order/createOrder).
        foreach (var segment in NonIdempotentPaths.All)
        {
            if (path.Contains(segment, StringComparison.OrdinalIgnoreCase))
                return true;
        }

        return false;
    }

    private static bool IsServerError(HttpResponseMessage response) =>
        (int)response.StatusCode >= 500;

    private async Task TryCancelOrderAsync(HttpRequestMessage originalRequest, string? requestBody)
    {
        try
        {
            var requestId = ExtractRequestId(requestBody);
            if (requestId is null)
                return;

            var cancelUri = BuildCancelUri(originalRequest);
            if (cancelUri is null)
                return;

            using var cancelRequest = new HttpRequestMessage(HttpMethod.Post, cancelUri);
            cancelRequest.Content = new StringContent(
                JsonSerializer.Serialize(new { originalRequestId = requestId }),
                Encoding.UTF8,
                "application/json");

            // Copy the auth header so the cancel request authenticates with the same token.
            if (originalRequest.Headers.Authorization is { } auth)
                cancelRequest.Headers.Authorization = auth;

            // Copy the partner-id header in case no PartnerIdHandler is registered.
            if (originalRequest.Headers.TryGetValues(MissionControlHeaders.PartnerId, out var partnerIds))
                cancelRequest.Headers.TryAddWithoutValidation(MissionControlHeaders.PartnerId, partnerIds);

            using var cts = new CancellationTokenSource(CancelTimeout);
            await base.SendAsync(cancelRequest, cts.Token).ConfigureAwait(false);
        }
        catch
        {
            // Best-effort: swallow failures so the original error propagates.
        }
    }

    private static Guid? ExtractRequestId(string? json)
    {
        if (string.IsNullOrEmpty(json))
            return null;

        try
        {
            using var doc = JsonDocument.Parse(json);
            if (doc.RootElement.TryGetProperty("requestId", out var prop) && prop.TryGetGuid(out var id))
                return id;
        }
        catch (JsonException)
        {
            // Malformed body — nothing to extract.
        }

        return null;
    }

    private static Uri? BuildCancelUri(HttpRequestMessage request)
    {
        if (request.RequestUri is null)
            return null;

        var path = request.RequestUri.AbsolutePath;

        // Find the "/Order/" segment and replace everything after it with "cancelOrder".
        var orderIndex = path.IndexOf("/Order/", StringComparison.OrdinalIgnoreCase);
        if (orderIndex < 0)
            return null;

        var newPath = string.Concat(path.AsSpan(0, orderIndex + "/Order/".Length), "cancelOrder");
        var builder = new UriBuilder(request.RequestUri) { Path = newPath, Query = string.Empty };
        return builder.Uri;
    }
}

File: node/src/client/autoCancelOrderHandler.ts
import type { RequestOption } from "@microsoft/kiota-abstractions";
import type { Middleware } from "@microsoft/kiota-http-fetchlibrary";
import { MissionControlHeaders } from "./missionControlHeaders";

const AUTO_CANCEL_PATH_PREFIXES = [
  "/Order/createOrder",
  "/Order/claimReservation",
  "/Order/reportOrder",
] as const;

const CANCEL_TIMEOUT_MS = 30_000;

/**
 * A Kiota middleware that automatically attempts to cancel an order when a request to one of
 * the order-mutation endpoints fails with a server error (5xx) or a network/connection error.
 *
 * Affected endpoints: `createOrder`, `claimReservation`, `reportOrder`.
 *
 * The cancellation attempt is best-effort. If it fails the original error is still propagated
 * to the caller unchanged.
 */
export class AutoCancelOrderHandler implements Middleware {
  next: Middleware | undefined;

  async execute(
    url: string,
    requestInit: RequestInit,
    requestOptions?: Record<string, RequestOption>,
  ): Promise<Response> {
    if (!isAutoCancelEndpoint(url)) {
      return this.next!.execute(url, requestInit, requestOptions);
    }

    // Buffer the request body before the inner handler consumes it.
    const requestBody =
      requestInit.body != null ? String(requestInit.body) : null;

    let response: Response;
    try {
      response = await this.next!.execute(url, requestInit, requestOptions);
    } catch (err) {
      // Network/connection error — attempt a best-effort cancel.
      await this.tryCancelOrderAsync(url, requestInit, requestBody);
      throw err;
    }

    if (isServerError(response)) {
      await this.tryCancelOrderAsync(url, requestInit, requestBody);
    }

    return response;
  }

  private async tryCancelOrderAsync(
    originalUrl: string,
    originalInit: RequestInit,
    requestBody: string | null,
  ): Promise<void> {
    try {
      const requestId = extractRequestId(requestBody);
      if (requestId == null) return;

      const cancelUrl = buildCancelUrl(originalUrl);
      if (cancelUrl == null) return;

      const headers = new Headers();
      headers.set("Content-Type", "application/json");

      // Copy auth and partner-id headers so the cancel request is properly authenticated
      // and attributed — the inner middleware handles retry and other cross-cutting concerns.
      const originalHeaders = new Headers(originalInit.headers);
      const auth = originalHeaders.get("Authorization");
      if (auth) headers.set("Authorization", auth);
      const partnerId = originalHeaders.get(MissionControlHeaders.PartnerId);
      if (partnerId) headers.set(MissionControlHeaders.PartnerId, partnerId);

      const cancelInit: RequestInit = {
        method: "POST",
        headers,
        body: JSON.stringify({ originalRequestId: requestId }),
        signal: AbortSignal.timeout(CANCEL_TIMEOUT_MS),
      };

      await this.next!.execute(cancelUrl, cancelInit, undefined);
    } catch {
      // Best-effort: swallow failures so the original error propagates.
    }
  }
}

function isAutoCancelEndpoint(url: string): boolean {
  try {
    const path = new URL(url).pathname;
    return AUTO_CANCEL_PATH_PREFIXES.some((prefix) =>
      path.toLowerCase().includes(prefix.toLowerCase()),
    );
  } catch {
    return false;
  }
}

function isServerError(response: Response): boolean {
  return response.status >= 500;
}

function extractRequestId(json: string | null): string | null {
  if (!json) return null;
  try {
    const parsed: unknown = JSON.parse(json);
    if (
      parsed != null &&
      typeof parsed === "object" &&
      "requestId" in parsed &&
      typeof (parsed as Record<string, unknown>).requestId === "string"
    ) {
      return (parsed as Record<string, unknown>).requestId as string;
    }
  } catch {
    // Malformed body — nothing to extract.
  }
  return null;
}

function buildCancelUrl(url: string): string | null {
  try {
    const parsed = new URL(url);
    const path = parsed.pathname;
    const orderIndex = path.toLowerCase().indexOf("/order/");
    if (orderIndex < 0) return null;

    parsed.pathname =
      path.slice(0, orderIndex + "/Order/".length) + "cancelOrder";
    parsed.search = "";
    return parsed.toString();
  } catch {
    return null;
  }
}
